### PR TITLE
Adding basic support for AMD SEV VMs

### PR DESCRIPTION
**Pull request by gabrielmougard**
Specification description: https://discuss.linuxcontainers.org/t/lxd-support-for-amd-sev-in-vm/16642 

- [ ] Change LXD API to return SEV information
- [x] Handling Diffie-Hellman key and session blob
- [x] SEV/SEV-ES feature check on host
- [x] QEMU flags to enable SEV/SEV-ES
- [x] Document SEV instance options and API extension

WIP: This PR needs to be tested on a Rome/Naples/Milan architecture

### CODE REVIEW

File: .sphinx/.wordlist.txt [hunk 1]
```diff
@@ -280,3 +281,9 @@ Zettabyte
 ZFS
 zpool
 zpools
+AMD
+SEV
+ES
+Diffie
+Hellman
+base64
```
Comment (stgraber) [inline .sphinx/.wordlist.txt:289]: @ru-fu can you go through this commit and see which ones actually deserve an exception and which ones just need quoting?

**Comment by stgraber** [comment]
To make feature detection actually work on hardware, I had to do:
```
diff --git a/lxd/instance/drivers/driver_qemu.go b/lxd/instance/drivers/driver_qemu.go
index 94b895f23..9d1be60ba 100644
--- a/lxd/instance/drivers/driver_qemu.go
+++ b/lxd/instance/drivers/driver_qemu.go
@@ -6397,6 +6397,7 @@ func (d *qemu) checkFeatures(hostArch int, qemuPath string) ([]string, error) {
 		"-nographic",
 		"-nodefaults",
 		"-no-user-config",
+		"-accel", "kvm",
 		"-chardev", fmt.Sprintf("socket,id=monitor,path=%s,server=on,wait=off", monitorPath.Name()),
 		"-mon", "chardev=monitor,mode=control",
 		"-machine", qemuMachineType(hostArch),
@@ -6520,7 +6521,7 @@ func (d *qemu) checkFeatures(hostArch int, qemuPath string) ([]string, error) {
 		if err != nil {
 			logger.Debug("Failed reading SEV capability during VM feature check", logger.Ctx{"err": err})
 		} else {
-			if string(sev) == "Y" {
+			if string(sev) == "Y\n" {
 				// Host supports SEV, check if QEMU supports it as well.
 				capabilities, err := monitor.SEVCapabilities()
 				if err != nil {
 
diff --git a/lxd/instance/drivers/qmp/commands.go b/lxd/instance/drivers/qmp/commands.go
index f73628019..ca512327d 100644
--- a/lxd/instance/drivers/qmp/commands.go
+++ b/lxd/instance/drivers/qmp/commands.go
@@ -694,20 +694,23 @@ func (m *Monitor) AddSecret(id string, secret string) error {
 // AMDSEVCapabilities represents the SEV capabilities of QEMU.
 type AMDSEVCapabilities struct {
 	PDH             string `json:"pdh"`               // Platform Diffie-Hellman key (base64-encoded)
-	CertChain       string `json:"cert_chain"`        // PDH certificate chain (base64-encoded)
-	CPU0Id          string `json:"cpu0_id"`           // Unique ID of CPU0 (base64-encoded)
+	CertChain       string `json:"cert-chain"`        // PDH certificate chain (base64-encoded)
+	CPU0Id          string `json:"cpu0-id"`           // Unique ID of CPU0 (base64-encoded)
 	CBitPos         int    `json:"cbitpos"`           // C-bit location in page table entry
-	ReducedPhysBits int    `json:"reduced_phys_bits"` // Number of physical address bit reduction when SEV is enabled
+	ReducedPhysBits int    `json:"reduced-phys-bits"` // Number of physical address bit reduction when SEV is enabled
 }
 
 // SEVCapabilities is used to get the SEV capabilities, and is supported on AMD X86 platforms only.
 func (m *Monitor) SEVCapabilities() (AMDSEVCapabilities, error) {
-	var resp AMDSEVCapabilities
+	// Prepare the response
+	var resp struct {
+		Return AMDSEVCapabilities `json:"return"`
+	}
 
-	err := m.run("query-sev-capability", nil, &resp)
+	err := m.run("query-sev-capabilities", nil, &resp)
 	if err != nil {
 		return AMDSEVCapabilities{}, fmt.Errorf("Failed querying SEV capability for QEMU: %w", err)
 	}
 
-	return resp, nil
+	return resp.Return, nil
 }
```

**Comment by stgraber** [comment]
Moving back to draft until we have this tested on a physical system and all items in the description covered.

**Comment by stgraber** [comment]
I've rebased and updated the commit structure to match what we usually expect for a LXD feature.
Going to be doing some testing next.

I'd still like to see the `lxd/instance/qemu` commit be split in half, first having a commit that adds the feature detection and then a second which implements security.sev.

**Comment by stgraber** [comment]
Needs a tweak to the AppArmor profile for QEMU.

**Comment by stgraber** [comment]
With the AppArmor profile altered, I'm getting:
```
root@infra06:~# lxc start test-vm
Error: Failed to run: forklimits limit=memlock:unlimited:unlimited fd=3 -- /snap/lxd/24615/bin/qemu-system-x86_64 -S -name test-vm -uuid 6e4dfd34-5754-40d6-8da2-3d35e501ae09 -daemonize -cpu host,hv_passthrough -nographic -serial chardev:console -nodefaults -no-user-config -sandbox on,obsolete=deny,elevateprivileges=allow,spawn=allow,resourcecontrol=deny -readconfig /var/snap/lxd/common/lxd/logs/cloud_test-vm/qemu.conf -spice unix=on,disable-ticketing=on,addr=/var/snap/lxd/common/lxd/logs/cloud_test-vm/qemu.spice -pidfile /var/snap/lxd/common/lxd/logs/cloud_test-vm/qemu.pid -D /var/snap/lxd/common/lxd/logs/cloud_test-vm/qemu.log -device vmgenid,guid=6e4dfd34-5754-40d6-8da2-3d35e501ae09 -smbios type=2,manufacturer=Canonical Ltd.,product=LXD -machine q35,confidential-guest-support=sev0 -object sev-guest,id=sev0,cbitpos=47,reduced-phys-bits=1,policy=0x1 -runas lxd: : Process exited with non-zero value 1
Try `lxc info --show-log test-vm` for more info
root@infra06:~# lxc info --show-log test-vm
Name: test-vm
Status: STOPPED
Type: virtual-machine
Architecture: x86_64
Location: infra06
Created: 2023/03/20 14:32 UTC
Last Used: 2023/03/20 14:33 UTC

Log:

qemu-system-x86_64:/var/snap/lxd/common/lxd/logs/cloud_test-vm/qemu.conf:231: iommu_platform=true is not supported by the device

```

**Comment by stgraber** [comment]
Line 231 is the qemu_config chardev at:
```
[device "dev-qemu_config-drive-virtio-fs"]
driver = "vhost-user-fs-pci"
bus = "qemu_pcie2"
addr = "00.1"
tag = "config"
chardev = "qemu_config"
```

**Comment by stgraber** [comment]
Disabling virtiofs with `migration.stateful=true`, VM then starts fine.

**Comment by stgraber** [comment]
@gabrielmougard please ping me on IRC once you've addressed the above and I'll test again to make sure SEV works.

File: lxd/instance/drivers/driver_qemu.go [hunk 2]
```diff
@@ -1313,6 +1313,136 @@ func (d *qemu) start(stateful bool, op *operationlock.InstanceOperation) error {
...
+							policy,
+						),
+					)
+				}
+
+				if shared.IsTrue(d.localConfig["security.sev.policy.es"]) {
+					if slices.Contains(features, "sev-es") {
+						addSEVCmd("0x5")
+					} else {
+						d.logger.Warn("You tried to enable AMD SEV-ES but the host doesn't support it, ignoring")
```
Comment (stgraber) [inline lxd/instance/drivers/driver_qemu.go:1386]: Should be `Host system doesn't support AMD SEV-ES` and this should be an error that prevents startup.

File: lxd/instance/drivers/driver_qemu.go [hunk 3]
```diff
@@ -1313,6 +1313,136 @@ func (d *qemu) start(stateful bool, op *operationlock.InstanceOperation) error {
...
+						"-object",
+						fmt.Sprintf(
+							"sev-guest,id=sev0,cbitpos=%d,reduced-phys-bits=%d,%spolicy=%s",
+							cbitpos,
+							reducedPhysBits,
+							optSessionFiles,
+							policy,
+						),
+					)
+				}
```
Comment (stgraber) [inline lxd/instance/drivers/driver_qemu.go]: We still need to turn that into qemu.conf

File: .sphinx/.wordlist.txt [hunk 4]
```diff
@@ -61,6 +63,7 @@ ECDSA
 EiB
 Eibit
 endian
+ES
```
Comment (stgraber) [inline .sphinx/.wordlist.txt:66]: @ru-fu happy with those?

Personally, I'm not convinced that `ES` and `SEV` are really suitable for inclusion.
And I'd prefer we add `Diffie-Hellman` rather than have it as two words, especially as I believe it is meant to be written with a hyphen in the first place.

File: lxd/apparmor/instance.go [hunk 5]
```diff
@@ -198,6 +198,13 @@ func instanceProfile(sysOS *sys.OS, inst instance) (string, error) {
 			execPath = execPathFull
 		}
 
+		// Check that /dev/sev exists to pass it to the AppArmor profile.
```
Comment (stgraber) [inline lxd/apparmor/instance.go:201]: I don't think the test is warranted, apparmor won't mind us listing a non-existent path in the profile, so let's keep it simple and just make it a one line addition to the profile.

File: lxd/instance/drivers/driver_qemu.go [hunk 6]
```diff
@@ -1371,34 +1372,21 @@ func (d *qemu) start(stateful bool, op *operationlock.InstanceOperation) error {
...
-						"-object",
-						fmt.Sprintf(
-							"sev-guest,id=sev0,cbitpos=%d,reduced-phys-bits=%d,%spolicy=%s",
-							cbitpos,
-							reducedPhysBits,
-							optSessionFiles,
-							policy,
-						),
-					)
+				sevOpts.cbitpos = cbitpos
```
Comment (stgraber) [inline lxd/instance/drivers/driver_qemu.go:1375]: This stuff should be merged into the original commit so we don't introduce logic that we remove two commits later in the same branch. This causes time wastage during review as we go from commit to commit and therefore leave unapplicable comments.

File: .sphinx/.wordlist.txt (hunk 4 above)
Comment (ru-fu) [reply .sphinx/.wordlist.txt:66]: If we want to introduce the abbreviations SEV and SEV-ES, we need to include them here or the spellchecker will complain. And since we're using them in the option names and the messages, I think we should introduce them.

For the hyphen, that's a limitation of the spellchecker - it treats a hyphen as a separator. Which means it will be looking for Diffie and Hellman separately. :(

So yes, happy with the update. :)

File: lxd/instance/drivers/driver_qemu.go [hunk 7]
```diff
@@ -7061,6 +7062,50 @@ func (d *qemu) checkFeatures(hostArch int, qemuPath string) ([]string, error) {
...
+	if hostArch == osarch.ARCH_64BIT_INTEL_X86 {
+		cmdline, err := os.ReadFile("/proc/cmdline")
+		if err != nil {
+			return nil, err
+		}
+
+		parts := strings.Split(string(cmdline), " ")
+
+		// Check if SME is enabled in the kernel command line.
+		if slices.IndexFunc(parts, func(p string) bool { return p == "mem_encrypt=on" }) != -1 {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:7212]: You can just use `shared.StringInSlice()` here I think.

File: lxd/instance/drivers/driver_qemu.go [hunk 8]
```diff
@@ -7061,6 +7062,50 @@ func (d *qemu) checkFeatures(hostArch int, qemuPath string) ([]string, error) {
...
+			features = append(features, "sme")
+		}
+
+		// Check if SEV/SEV-ES are enabled
+		sev, err := os.ReadFile("/sys/module/kvm_amd/parameters/sev")
+		if err != nil {
+			logger.Debug("Failed reading SEV capability during VM feature check", logger.Ctx{"err": err})
+		} else {
+			if string(sev) == "Y\n" {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:7221]: Is there a reason you used a separate indented `if` block here rather than make the line above and `else if`?

File: lxd/instance/drivers/driver_qemu.go [hunk 9]
```diff
@@ -1313,6 +1313,136 @@ func (d *qemu) start(stateful bool, op *operationlock.InstanceOperation) error {
...
+	if shared.IsTrue(d.localConfig["security.sev"]) {
+		if d.architecture == osarch.ARCH_64BIT_INTEL_X86 {
+			// Get the QEMU features to check if AMD SEV is supported.
+			features, err := d.checkFeatures(d.architecture, qemuPath)
+			if err != nil {
+				op.Done(err)
+				return err
+			}
+
+			if slices.Contains(features, "sme") && slices.Contains(features, "sev") {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1290]: You can use shared.StringInSlice() here. Save another import dependency.

File: lxd/instance/drivers/driver_qemu.go [hunk 10]
```diff
@@ -7061,6 +7062,50 @@ func (d *qemu) checkFeatures(hostArch int, qemuPath string) ([]string, error) {
...
+
+		// Check if SME is enabled in the kernel command line.
+		if slices.IndexFunc(parts, func(p string) bool { return p == "mem_encrypt=on" }) != -1 {
+			features = append(features, "sme")
+		}
+
+		// Check if SEV/SEV-ES are enabled
+		sev, err := os.ReadFile("/sys/module/kvm_amd/parameters/sev")
+		if err != nil {
+			logger.Debug("Failed reading SEV capability during VM feature check", logger.Ctx{"err": err})
```
[the remark this answers is not in the document]
Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:7219]: Ah, no, this is checkFeatures, this is called outside of the instance context, so no contextualised logger, its fine as it is then.

[the remark this answers is not in the document]
Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:7219]: `	if err != nil && !os.IsNotExist(err) {`

File: lxd/instance/drivers/driver_qemu.go [hunk 11]
```diff
@@ -7061,6 +7062,50 @@ func (d *qemu) checkFeatures(hostArch int, qemuPath string) ([]string, error) {
...
+		} else {
+			if string(sev) == "Y\n" {
+				// Host supports SEV, check if QEMU supports it as well.
+				capabilities, err := monitor.SEVCapabilities()
+				if err != nil {
+					logger.Debug("Failed querying SEV capability during VM feature check", logger.Ctx{"err": err})
+				} else {
+					features = append(features, "sev")
+					features = append(features, fmt.Sprintf("sev.cbitpos=%d", capabilities.CBitPos))
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:7228]: Hrm, I'm not really enjoying the `feature=value` encoding technique we are introducing here, as it then requires decoding elsewhere. This is something that (with hindsight) we also mistakenly fell into with the RunConfig shared between device driver and instance driver, and ended up having to do use custom (undocumented) encoding of various things.

Because these values are just used internally, I suggest we change the features slice into a `map[string]any` (this would need to be a series of commits earlier in this PR or a separate PR before introducing the SEV concept). 

Then for features that have no additional info the properties, it can just store the name of the feature and nil as the value, and then for features that do have additional properties they can be stored as the actual data type that `monitor.SEVCapabilites` returns.

Checking for a feature and then accessing the feature options then becomes easy as one can do something like this:

```
featureOpts, found := features["foo"]
if found {
   fooFeatureOpts, ok := featureOpts.(dataType)
  if !ok {
      // Unexpected invalid data type.
  }
}
```

Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:7228]: That way we don't have to do any manual encoding/decoding, and we benefit from Go's type checking, as well as being able to more easily reference where these data types are used.

File: lxd/instance/drivers/driver_qemu.go [hunk 12]
```diff
@@ -1313,6 +1313,136 @@ func (d *qemu) start(stateful bool, op *operationlock.InstanceOperation) error {
 		qemuCmd = append(qemuCmd, "-smbios", "type=2,manufacturer=Canonical Ltd.,product=LXD")
 	}
 
+	// If user has requested AMD SEV, check if supported and add to QEMU command.
+	if shared.IsTrue(d.localConfig["security.sev"]) {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1281]: Lets move this whole bit into a separate function called from here as `start()` is already too long IMHO.
Its getting hard to follow what the function is doing when you're not concerned with the nitty gritty detail of each step.

**Comment by gabrielmougard** [comment]
@tomponline @stgraber I reworked the overall commits structure taking into account your updates. Tell me if I'm forgetting anything.

**Comment by gabrielmougard** [comment]
> Hrm, I'm not really enjoying the `feature=value` encoding technique we are introducing here, as it then requires decoding elsewhere. This is something that (with hindsight) we also mistakenly fell into with the RunConfig shared between device driver and instance driver, and ended up having to do use custom (undocumented) encoding of various things.
> 
> Because these values are just used internally, I suggest we change the features slice into a `map[string]any` (this would need to be a series of commits earlier in this PR or a separate PR before introducing the SEV concept).
> 
> Then for features that have no additional info the properties, it can just store the name of the feature and nil as the value, and then for features that do have additional properties they can be stored as the actual data type that `monitor.SEVCapabilites` returns.
> 
> Checking for a feature and then accessing the feature options then becomes easy as one can do something like this:
> 
> ```go
> featureOpts, found := features["foo"]
> if found {
>    fooFeatureOpts, ok := featureOpts.(dataType)
>    if !ok {
>       // Unexpected invalid data type.
>    }
> }
> ```

@tomponline I'm thinking about doing it in the same PR. The change is coming.

File: lxd/instance/drivers/driver_qemu.go [hunk 13]
```diff
@@ -6895,7 +6896,7 @@ func (d *qemu) devlxdEventSend(eventType string, eventMessage map[string]any) er
 func (d *qemu) Info() instance.Info {
 	data := instance.Info{
 		Name:     "qemu",
-		Features: []string{},
+		Features: make(map[string]any),
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:7021]: What does the https://github.com/lxc/lxd/blob/master/lxd/instance/drivers/load.go#L176 look like in the logs now with this change? I'm wondering if we need to alter that line to just log the feature names, rather than their specific options.

File: lxd/instance/drivers/driver_qemu.go [hunk 14]
```diff
@@ -7075,6 +7075,48 @@ func (d *qemu) checkFeatures(hostArch int, qemuPath string) (map[string]any, err
...
+		sev, err := os.ReadFile("/sys/module/kvm_amd/parameters/sev")
+		if err != nil && !os.IsNotExist(err) {
+			return nil, err
+		} else if string(sev) == "Y\n" {
+			// Host supports SEV, check if QEMU supports it as well.
+			capabilities, err := monitor.SEVCapabilities()
+			if err != nil {
+				logger.Debug("Failed querying SEV capability during VM feature check", logger.Ctx{"err": err})
+			} else {
+				features["sev"] = struct{}{}
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:7225]: I was thinking we should actually store the `capabilities` var in `features["sev"]`, rather than split the "sev" feature into multiple sub-features, when really its just properties of the "sev" feature.

File: lxd/instance/drivers/driver_qemu.go [hunk 15]
```diff
@@ -7075,6 +7075,48 @@ func (d *qemu) checkFeatures(hostArch int, qemuPath string) (map[string]any, err
...
+				features["sev.cbitpos"] = capabilities.CBitPos
+				features["sev.reduced_phys_bits"] = capabilities.ReducedPhysBits
+
+				// If SEV is enabled on host and supported by QEMU,
+				// check if the SEV-ES extension is enabled.
+				sevES, err := os.ReadFile("/sys/module/kvm_amd/parameters/sev_es")
+				if err != nil {
+					logger.Debug("Failed querying SEV-ES capability during VM feature check", logger.Ctx{"err": err})
+				} else {
+					if string(sevES) == "Y" {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:7235]: Above you tested for `Y\n` when using os.ReadFile and here you're only testing for `Y`, is this a bug?

Perhaps for readability and safety we should pass both `/sys/module/kvm_amd/parameters/sev` and `/sys/module/kvm_amd/parameters/sev_es` through `strings.TrimSpace()` before comparison.

File: lxd/instance/drivers/driver_qemu.go [hunk 16]
```diff
@@ -1221,7 +1221,17 @@ func (d *qemu) start(stateful bool, op *operationlock.InstanceOperation) error {
 	// Define a set of files to open and pass their file descriptors to qemu command.
 	fdFiles := make([]*os.File, 0)
 
-	confFile, monHooks, err := d.generateQemuConfigFile(mountInfo, qemuBus, devConfs, &fdFiles)
+	// If user has requested AMD SEV, check if supported and add to QEMU config.
+	sevOpts := &qemuSevOpts{}
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1225]: Is it possible to move the call to d.setupSEV into d.generateQemuConfigFile, like you did with addVmgenDeviceConfig?

File: lxd/instance/drivers/driver_qemu.go [hunk 17]
```diff
@@ -1600,6 +1600,105 @@ func (d *qemu) start(stateful bool, op *operationlock.InstanceOperation) error {
...
+	_, smeFound := features["sme"]
+	sev, sevFound := features["sev"]
+	if !smeFound || !sevFound {
+		return errors.New("AMD SEV is not supported by the host")
+	}
+
+	// Get the SEV guest `cbitpos` and `reducedPhysBits`.
+	sevCapabilities, ok := sev.(qmp.AMDSEVCapabilities)
+	if !ok {
+		return errors.New("Failed to get the guest `sev` capabilities")
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1628]: Just a nit but we use double quotes in error messages, so this should be:

```go
		return errors.New(`Failed to get the guest "sev" capabilities`)
```

File: lxd/instance/drivers/driver_qemu.go [hunk 18]
```diff
@@ -1600,6 +1600,105 @@ func (d *qemu) start(stateful bool, op *operationlock.InstanceOperation) error {
...
+	}
+
+	cbitpos := sevCapabilities.CBitPos
+	reducedPhysBits := sevCapabilities.ReducedPhysBits
+
+	// Write user's dh-cert and session-data to file descriptors.
+	var dhCertFD, sessionDataFD int
+	var dhCertFDOk, sessionDataFDOk bool
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1636]: Do we need these vars, if the associated FD variables are >0 then they are OK I believe?

File: lxd/instance/drivers/driver_qemu.go [hunk 19]
```diff
@@ -2721,6 +2820,26 @@ func (d *qemu) generateQemuConfigFile(mountInfo *storagePools.MountInfo, busName
 
 	cfg = append(cfg, qemuDriveConfig(&driveConfig9pOpts)...)
 
+	// If user has requested AMD SEV, check if supported and add to QEMU config.
+	sevOpts := &qemuSevOpts{}
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:2824]: I can't see why sevOpts is defined outside of the if block below, we should try and limit the scope of variables wherever possible to avoid confusion. Also why don't we modify d.setupSev to return the `qemuSevOpts` struct, rather than pass a pointer to it to be modified? Then you can directly assign it to `sevOpts` here. What do you think?

Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:2824]: Looks like static analysis has picked up some issues here as well.

Comment (gabrielmougard) [reply lxd/instance/drivers/driver_qemu.go:2824]: thanks ! Taking care of all that

File: lxd/instance/drivers/driver_qemu.go [hunk 20]
```diff
@@ -1600,6 +1600,105 @@ func (d *qemu) start(stateful bool, op *operationlock.InstanceOperation) error {
...
+			return err
+		}
+
+		_, err = dhCert.WriteString(d.localConfig["security.sev.session.dh"])
+		if err != nil {
+			return err
+		}
+
+		dhCertFD = d.addFileDescriptor(fdFiles, dhCert)
+		defer dhCert.Close()
```
[the remark this answers is not in the document]
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_qemu.go:1654]: I'm deleting the `defer dhCert.Close()` and defer `sessionData.Close()` lines but as you can see line 1411, the files are closed with a deferred call. Plus, adding the deferred calls line 1223 would not contain our files that are being added after so I think this should be good.

File: lxd/instance/drivers/driver_qemu.go [hunk 21]
```diff
@@ -1600,6 +1600,105 @@ func (d *qemu) start(stateful bool, op *operationlock.InstanceOperation) error {
...
+	sevOpts.reducedPhysBits = reducedPhysBits
+	if dhCertFDOk && sessionDataFDOk {
+		sevOpts.dhCertFD = fmt.Sprintf("/proc/fd/%d", dhCertFD)
+		sevOpts.sessionDataFD = fmt.Sprintf("/proc/fd/%d", sessionDataFD)
+	}
+
+	if shared.IsTrue(d.localConfig["security.sev.policy.es"]) {
+		_, sevES := features["sev-es"]
+		if sevES {
+			sevOpts.policy = "0x5"
```
[the remark this answers is not in the document]
Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:1687]: Thanks!

File: lxd/instance/drivers/driver_qemu.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go]: @gabrielmougard mmm, ok, but we still need to have the cleanup code I posted above (in the place I suggested) and then remove the existing defer cleanup code on line 1411. As otherwise if the VM start functions fails to get to line 1411 then the file handles will be left open and leak.

> Plus, adding the deferred calls line 1223 would not contain our files that are being added after so I think this should be good.

Thats not quite right. Whilst true that the file handles are not in fdFiles at the time of the proposed defer, notice that I am not suggesting we use the existing invocation of:

```go
// Ensure passed files are closed after qemu has started.
for _, file := range fdFiles {
    defer func(file *os.File) { _ = file.Close() }(file)
}
```

Which, as you say, would not actually do anything because fdFiles is empty.

I am proposing using:

```go
// Ensure passed files are closed after start has returned (either because qemu has started or on error).
defer func() {
	for _, file := range fdFiles {
		_ = file.Close()
	}
}()
```

Which says instead: "at the end of the function, loop through fdFiles and close them".

**Review by tomponline — CHANGES_REQUESTED** [review]
Its looking good to me, just need to sort out those file handle cleanups with the defer, and then I'll let @stgraber take it for a spin to check it actually works before we merge it. Thanks

File: lxd/instance/drivers/driver_qemu.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (gabrielmougard) [reply lxd/instance/drivers/driver_qemu.go]: oh right ! Very subtle approach indeed ! I'm adding this bit then ;)

File: lxd/instance/drivers/driver_qemu.go [hunk 22]
```diff
@@ -1221,6 +1221,13 @@ func (d *qemu) start(stateful bool, op *operationlock.InstanceOperation) error {
 	// Define a set of files to open and pass their file descriptors to qemu command.
 	fdFiles := make([]*os.File, 0)
 
+	// Ensure passed files are closed after start has returned (either because qemu has started or on error).
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1224]: This change and the removal of the defer below should be a separate commit as its not really related to SEV change and is a good candidate for back porting to LTS branch (where new features are not backported), so we don't want to tightly couple it to the SEV change.

**Review by tomponline — APPROVED** [review]
LGTM!

Over to @stgraber to do a final test before merging.

**Comment by tomponline** [comment]
Good catch @stgraber

**Comment by stgraber** [comment]
```
diff --git a/lxd/instance/drivers/driver_qemu.go b/lxd/instance/drivers/driver_qemu.go
index cc118e90b..fa8949586 100644
--- a/lxd/instance/drivers/driver_qemu.go
+++ b/lxd/instance/drivers/driver_qemu.go
@@ -2827,9 +2827,9 @@ func (d *qemu) generateQemuConfigFile(mountInfo *storagePools.MountInfo, busName
                }
 
                if sevOpts != nil {
-                       for _, section := range cfg {
-                               if section.name == "machine" {
-                                       section.entries = append(section.entries, cfgEntry{"memory-encryption", "sev0"})
+                       for i, _ := range cfg {
+                               if cfg[i].name == "machine" {
+                                       cfg[i].entries = append(cfg[i].entries, cfgEntry{"memory-encryption", "sev0"})
                                        break
                                }
                        }
```

**Comment by stgraber** [comment]
Needs to be `/proc/self/fd` not `/proc/fd`

**Comment by stgraber** [comment]
`/tmp/lxd_sev_* r,` in apparmor rule to allow reading the fds
